index.js

```diff
@@ -119,16 +99,80 @@
   //  $$type :: String
   var $$type = '@@type';
 
-  //  type :: Any -> String
+  //  RPARSE :: RegExp
+  var RPARSE = /^(?:((?!\.|_)@?[0-9a-zA-Z./_%-]{1,214}):)?([^]+?)(?:@(\d+))?$/;
+
+  //       TypeIdentifier :: (Nullable String, String, Number)
+  //                      -> TypeIdentifier
+  function TypeIdentifier(namespace, name, version) {
+    return {
+      namespace: namespace,
+      name: name,
+      version: version
+    };
+  }
+
+  //       parseTypeIdentifier :: String -> TypeIdentifier
+  function parseTypeIdentifier(s) {
+    if (s === '') return TypeIdentifier(null, s, 1);
+    var parsed = RPARSE.exec(s);
```

Comment:
We could be more forgiving by switching the order of the lines above:

```javascript
var parsed = RPARSE.exec(s);
if (parsed == null) return TypeIdentifier(null, s, 1);
```

We'd then handle all invalid inputs consistently.